[WIP] event transformations

wozniakjan opened the pull request:
Initial implementation of event transformations of logs from the eventrouter, configurable by setting proper formatter (`k8s_json_file` or `k8s_journal`)

Example:
```
<filter **>
  @type viaq_data_model
  ..
  <formatter>
    type k8s_journal
    tag "kubernetes.**"
  </formatter>
  ..
</filter>
```

There is an optional argument to keep both formatters working without transforming event logs:
```
process_kubernetes_events   false
```

Transforms following fields:
```
kubernetes.event                                 <- event
pipeline_metadata.collector.original_raw_message <- message
message                                          <- event.message
@dest_time_name                                  <- event.metadata.creationTimestamp
```

cc: @richm , @portante

Commit:
Event transformations

Initial implementation of event transformations of logs from the eventrouter,
configurable by setting proper formatter (k8s_json_file or k8s_journal)

Example:
<filter **>
  @type viaq_data_model
  ..
  <formatter>
    type k8s_journal
    tag "kubernetes.**"
  </formatter>
  ..
</filter>

There is an optional argument to keep both formatters working without transforming
event logs:
process_kubernetes_events false

Transforms following fields:
kubernetes.event                                 <- event
pipeline_metadata.collector.original_raw_message <- message
message                                          <- event.message
@dest_time_name                                  <- event.metadata.creationTimestamp

wozniakjan commented:
I am trying to figure out why are the tests failing, the changes should be disabled by default so they shouldn't affect the current tests, but they appear to.

richm commented on lib/fluent/plugin/filter_viaq_data_model.rb:
```diff
@@ -90,6 +90,9 @@ class ViaqDataModelFilter < Filter
     config_param :dest_time_name, :string, default: '@timestamp'
 
+    desc 'Use event transformations, available options ["none", "eventrouter"]'
+    config_param :event_trans, :string, default: 'none'
```
If these are an enumerated list of options, use the `enum` type like this: https://github.com/ViaQ/fluent-plugin-viaq_data_model/blob/master/lib/fluent/plugin/filter_viaq_data_model.rb#L105

richm commented on lib/fluent/plugin/filter_viaq_data_model.rb:
```diff
@@ -157,6 +160,13 @@ def configure(conf)
+        log.info("Set event transformation: "+@event_trans.to_s)
+      else
+        log.warn("Unknown type of event transformation "+@event_trans.to_s+" allowed only "+allowed_trans_options.to_s)
+      end
```
Use `enum` type

richm commented on lib/fluent/plugin/filter_viaq_data_model.rb:
```diff
@@ -395,6 +427,7 @@ def filter(tag, time, record)
           log.error("output #{time} #{tag} #{record}")
         end
       end
+      transform_event(record)
```
move this up between `check_for_match_and_format` and `add_pipeline_metadata`

Also - this will be run against every tag.  I think you should add a `event_trans_tag` parameter which takes the tag(s) to match, as was done for the formatters and elasticsearch_index_names.

wozniakjan replied:
One of the transformations I wanted to do was:
```
pipeline_metadata.collector.original_raw_message <- message
```
and `add_pipeline_metadata()` create the field `pipeline_metadata.collector`, wouldn't we want to do the event transformation after calling `add_pipeline_metadata()` or should I implement this logic in `add_pipeline_metadata()`?

Similar with this transformation
```
@dest_time_name <- event.metadata.creationTimestamp
```
if this happens before [the original timestamp transformation](https://github.com/wozniakjan/fluent-plugin-viaq_data_model/blob/97f327dc55dd1a00652cf0271d01d345331c5624/lib/fluent/plugin/filter_viaq_data_model.rb#L426-L429), it will get lost. Or would you recommend not moving the creationTimestamp?

richm replied:
Ah, I see.  Then change `add_pipeline_metadata` as you need - add a parameter for `original_raw_message`, etc.

richm commented on lib/fluent/plugin/filter_viaq_data_model.rb:
```diff
@@ -157,6 +160,13 @@ def configure(conf)
+      @event_trans = @event_trans.to_sym
+      allowed_trans_options = [:none, :eventrouter]
+      if allowed_trans_options.include?(@event_trans)
+        log.info("Set event transformation: "+@event_trans.to_s)
```
running fluentd with `trace` level logging will dump all of the config.
If you want to check the config.  At any rate, if you use an `enum` for `event_trans`, all of this code goes away.

portante commented on lib/fluent/plugin/filter_viaq_data_model.rb:
```diff
@@ -351,6 +368,35 @@ def add_elasticsearch_index_name_field(tag, time, record)
+          if record["event"].has_key?("message")
+            record["pipeline_metadata"] = {} unless record.has_key?("pipeline_metadata")
+            record["pipeline_metadata"]["collector"] = {} unless record["pipeline_metadata"].has_key?("collector")
+            record["pipeline_metadata"]["collector"]["original_raw_message"] = record["message"]
```
Is this already being performed in `add_pipeline_metadata`?

richm replied:
yes

wozniakjan replied:
Is there any other `add_pipeline_metadata` than this one?
https://github.com/ViaQ/fluent-plugin-viaq_data_model/blob/cc806f964ed937ed84c842c498a0913a6a1a2470/lib/fluent/plugin/filter_viaq_data_model.rb#L281-L290

I can't seem to find it there and I am not sure what am I missing. Could you please explain?

richm replied:
What we mean is that we already have a function to add pipeline metadata, and you are duplicating that method in your code.

wozniakjan replied:
Oh, ok. Got it, thanks! :)

portante commented on lib/fluent/plugin/filter_viaq_data_model.rb:
```diff
@@ -90,6 +90,18 @@ class ViaqDataModelFilter < Filter
     desc 'Name of destination timestamp field'
     config_param :dest_time_name, :string, default: '@timestamp'
 
+    # <event_trans>
```
Why do we call this `event_trans`?  Are you planning on having more transformations than just from `eventrouter`?  Perhaps this framework could be used to transform events from a docker daemon?

If we are not going to go this route, then having the extra `type` specifier seems redundant with the `event_trans` name itself.

richm replied:
Then this could even be another `formatter`
```
<formatter>
  type k8s_event
  tag "kubernetes.event.**"
</formatter>
```
Except that once a formatter hits a tag, it will skip all others.  You would need to add some other discriminator, other than just the tag, unless you could somehow guarantee a specific tag pattern for k8s events.  For example, in a rewrite-tag filter, you could do something like
```
  rewriteruleN some_field_specific_to_k8s_events some_value_specific_to_k8s_events kubernetes.event
```
If there were such a field or such a value.

wozniakjan replied:
@portante you are correct, the plan is to perform transformations from docker events as well once we have them

richm commented on lib/fluent/plugin/filter_viaq_data_model.rb:
```diff
@@ -395,6 +441,7 @@ def filter(tag, time, record)
           log.error("output #{time} #{tag} #{record}")
         end
       end
+      transform_event(tag, record)
```
This should be much higher in the code, just before `add_pipeline_metadata`, unless you rewrite this to work as a `formatter`

wozniakjan replied:
I have commented on your recommendation previously here https://github.com/ViaQ/fluent-plugin-viaq_data_model/pull/9#discussion_r137520308

In my testing, if I did what you recommended and tested that, neither of these two transformations were performed due to reasons explained there.
```
pipeline_metadata.collector.original_raw_message <- message
@dest_time_name                                  <- event.metadata.creationTimestamp
```
Could you please advise what solution do you recommend to my concerns?

richm replied:
Sorry, I missed that earlier comment.  I suggest changing the `add_pipeline_metadata` function to add whatever you need.

richm commented on lib/fluent/plugin/filter_viaq_data_model.rb:
```diff
@@ -90,6 +90,18 @@ class ViaqDataModelFilter < Filter
+      desc 'Process records with this tag pattern'
+      config_param :tag, :string
+      desc 'Type of event transformation'
+      config_param :type, :enum, list: [:none, :eventrouter]
```
What is the purpose of having a `:none` type?

wozniakjan replied:
I had it here as a placeholder for docker event transformations

richm commented:
I think kubernetes events are just a special case of kubernetes log records, and should be handled as such.  That is - kubernetes events will be exactly the same as kubernetes log records, except that they will have an additional `event` field.
We could always process them - call the function `transform_eventrouter` from `process_k8s_json_file_fields`, and in the `when :k8s_journal` block at the end of `process_journal_fields`.  You can also do the `@timestamp` addition in `transform_eventrouter` too.
You could also handle the pipeline_metadata there too:
```
    ((record['pipeline_metadata'] ||= {})[pipeline_type] ||= {})["original_raw_message"] = record["message"]
    record["message"] = record["kubernetes"]["event"].delete("message") 
```

richm commented:
If we think that there will be some cases where the user will want to disable the extra processing for kubernetes events, and just process them as regular kubernetes log entries, then add a configuration parameter `process_kubernetes_events` bool with a default value of `true`, and have `transform_eventrouter` check for `@process_kubernetes_events`

wozniakjan commented:
> You can also do the @timestamp addition in transform_eventrouter too. You could also handle the pipeline_metadata there too:

But then we have the same problem described here https://github.com/ViaQ/fluent-plugin-viaq_data_model/pull/9#discussion_r137520308. 

That [this code](https://github.com/viaq/fluent-plugin-viaq_data_model/blob/97f327dc55dd1a00652cf0271d01d345331c5624/lib/fluent/plugin/filter_viaq_data_model.rb#L426-L429) will always overwrite the `@timestamp` and [this code](https://github.com/viaq/fluent-plugin-viaq_data_model/blob/97f327dc55dd1a00652cf0271d01d345331c5624/lib/fluent/plugin/filter_viaq_data_model.rb#L298-L307) will always overwrite the `original_raw_message` stored in `pipeline_metadata` because `transform_eventrouter` is called between `check_for_match_and_format` and `add_pipeline_metadata`.

I was trying to follow this [advice](https://github.com/ViaQ/fluent-plugin-viaq_data_model/pull/9#discussion_r137798107) but I am getting confused how should it work with putting everything back to the `transform_eventrouter` method

richm commented:
Sorry, I've been confusing you.
change the `transform_eventrouter` method like this:
```
    def transform_eventrouter(tag, record)
      if record.has_key?("event")
        if record.has_key?("verb")
          record["event"]["verb"] = record.delete("verb")
        end
        record["kubernetes"] = {} unless record.has_key?("kubernetes")
        record["kubernetes"]["event"] = record.delete("event")
        if record.has_key?('message')
          ((record['pipeline_metadata'] ||= {})[@pipeline_type.to_s] ||= {})['original_raw_message'] = record['message']
        end
        record['message'] = record["kubernetes"]["event"].delete("message")
        record['time'] = record["kubernetes"]["event"]["metadata"].delete("creationTimestamp") 
      end
    end
```
call this at the end of `process_k8s_json_file_fields`, and at the end of the `when :k8s_journal` case in `process_journal_fields`

Finally, change `add_pipeline_metadata` like this:
```
    def add_pipeline_metadata (tag, time, record)
      record['pipeline_metadata'] = {} unless record.has_key?('pipeline_metadata')
      pipeline_type = @pipeline_type.to_s
      # this will catch the case where pipeline_type doesn't exist, or is not a Hash
      record['pipeline_metadata'][pipeline_type] = {} unless record['pipeline_metadata'][pipeline_type].respond_to?(:fetch)
      record['pipeline_metadata'][pipeline_type]['ipaddr4'] = @ipaddr4
      record['pipeline_metadata'][pipeline_type]['ipaddr6'] = @ipaddr6
      record['pipeline_metadata'][pipeline_type]['inputname'] = 'fluent-plugin-systemd'
      record['pipeline_metadata'][pipeline_type]['name'] = 'fluentd'
      record['pipeline_metadata'][pipeline_type]['received_at'] =Time.now.utc.to_datetime.rfc3339(6),
      record['pipeline_metadata'][pipeline_type]['version'] = @pipeline_version
    end
```

That way, `record['time']` will be converted in to `record['@timestamp']`, preserving the value you set, and `record['pipeline_metadata'][@pipeline_type.to_s][`original_raw_message']` will also be preserved.

richm commented:
use `key?` instead of `has_key?`

wozniakjan commented:
@richm thank you very much for helping me to get this right! I think I finally got what you were trying to explain and updated both PRs, this and https://github.com/openshift/origin-aggregated-logging/pull/573.

I dropped the redundant `event_trans` configuration and now it is by default turned on with configuration of a proper `formatter`. It is rebased on master (post 0.0.9) as I noticed there were changes in timestamp processing.

Also as you suggested, there is optional argument `process_kubernetes_events` set to true by default. And when eventrouter is not deployed, auto-set to false.

richm commented:
looks good!
Next step is to add some unit tests for events.

Commit:
Unit tests for event transformations

wozniakjan commented:
I added two unit tests with `journal`. I didn't find any unit tests using `json_file` formatter, should I add some as well?

wozniakjan commented:
I added three basic tests, two for `journal` (with event processing turned on/off) and one for `json_file` (with event processing on)
